### PR TITLE
Feature/api acl access

### PR DESCRIPTION
**Pull request by daniel-samson**
## Description

API v8 does not check to see if user has permission to perform actions:

| API Action        |  ACL Action|
| ------------- | -----:|
| List      | list |
| Create |   edit |
| Read   |  view |
| Update|    edit |
| Delete |    delete |

## Motivation and Context

Restricts access to the data in SuiteCRM which is being accessed from the REST API

## How To Test This

- set up OAuth2
- set up a user account with a role
- test access to various modules

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [x] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

### CODE REVIEW

File: lib/API/OAuth2/Exception/GrantTypeNotAllowedForClient.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+class GrantTypeNotAllowedForClient extends OAuth2
+{
+    /**
+     * GrantTypeNotAllowedForClient constructor.
+     * @param string $message API Exception "$message"
+     * @param int $code
+     * @param $previous
+     */
+    public function __construct($message = '', $code = ExceptionCode::APPLICATION_UNHANDLED_BEHAVIOUR, $previous = null)
```
Comment (zoltankocsardi) [inline lib/API/OAuth2/Exception/GrantTypeNotAllowedForClient.php]: Why is that?

File: lib/API/OAuth2/Exception/OAuth2.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+ * display the words "Powered by SugarCRM" and "Supercharged by SuiteCRM".
+ */
+
+namespace SuiteCRM\API\OAuth2\Exception;
+
+use Psr\Log\LogLevel;
+use SuiteCRM\Enumerator\ExceptionCode;
+use \SuiteCRM\Exception\Exception;
+
+class OAuth2 extends Exception
```
Comment (zoltankocsardi) [inline lib/API/OAuth2/Exception/OAuth2.php]: If this is an exception, please, do not call this OAuth2

File: lib/API/OAuth2/Exception/OAuth2.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+{
+    /**
+     * OAuth2 constructor.
+     * @param string $message API Exception "$message"
+     * @param int $code
+     * @param $previous
+     */
+    public function __construct($message = '', $code = ExceptionCode::APPLICATION_UNHANDLED_BEHAVIOUR, $previous = null)
```
Comment (zoltankocsardi) [inline lib/API/OAuth2/Exception/OAuth2.php]: Again, why is that?

File: lib/API/OAuth2/Exception/OAuth2.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    /**
+     * Gives addition details to what caused the exception
+     * @see ApiController::generateJsonApiExceptionResponse()
+     * @return string
+     */
+    public function getDetail()
+    {
+        return 'SuiteCRM OAuth 2 Server has encountered an exception which has not been handled';
```
Comment (zoltankocsardi) [inline lib/API/OAuth2/Exception/OAuth2.php]: This ain't useful...

File: lib/API/v8/Controller/ModuleController.php [hunk 5]
```diff
@@ -436,9 +436,18 @@ public function createModuleRecord(Request $req, Response $res, array $args)
...
+        if (
+            isset($body['data']['id'])
+            && !empty($body['data']['id'])
+        ) {
+
+            $beanID = $body['data']['id'];
             $bean = \BeanFactory::getBean($moduleName, $beanID);
 
+            if (!$bean->ACLAccess('save')) {
+                throw new NotAllowed();
```
Comment (zoltankocsardi) [inline lib/API/v8/Controller/ModuleController.php]: That's  cool, but why isn't allowed, you said nothing.

File: lib/API/v8/Controller/ModuleController.php [hunk 6]
```diff
@@ -465,6 +474,10 @@ public function createModuleRecord(Request $req, Response $res, array $args)
             ->fromJsonApiRequest($body['data'])
             ->toSugarBean();
 
+        if (!$sugarBean->ACLAccess('save')) {
+            throw new NotAllowed();
```
Comment (zoltankocsardi) [inline lib/API/v8/Controller/ModuleController.php]: Same here.

File: lib/API/v8/Controller/ModuleController.php [hunk 7]
```diff
@@ -537,6 +550,10 @@ public function getModuleRecord(Request $req, Response $res, array $args)
             throw $exception;
         }
 
+        if (!$sugarBean->ACLAccess('view')) {
+            throw new NotAllowed();
```
Comment (zoltankocsardi) [inline lib/API/v8/Controller/ModuleController.php]: Same here. Okay, I give up, I'll leave a comment outside.

File: lib/API/OAuth2/Repositories/ClientRepository.php [hunk 8]
```diff
@@ -60,7 +61,7 @@ public function getClientEntity($clientIdentifier, $grantType, $clientSecret = n
         }
 
         if($client->allowed_grant_type !== $grantType) {
-            throw OAuthServerException::grantTypeNotAllowedForClient();
+            throw new GrantTypeNotAllowedForClient();
```
Comment (zoltankocsardi) [inline lib/API/OAuth2/Repositories/ClientRepository.php]: In this case, GrantTypeNotAllowedForClient looks completely unnecessary to me...

File: lib/API/v8/Controller/ModuleController.php [hunk 9]
```diff
@@ -436,9 +436,18 @@ public function createModuleRecord(Request $req, Response $res, array $args)
...
         }
 
-        if (!empty($beanID = $body['data']['id'])) {
+        if (
+            isset($body['data']['id'])
+            && !empty($body['data']['id'])
+        ) {
+
+            $beanID = $body['data']['id'];
             $bean = \BeanFactory::getBean($moduleName, $beanID);
```
Comment (zoltankocsardi) [inline lib/API/v8/Controller/ModuleController.php]: If you are in the middle of writing exceptions, maybe you can add some new exceptions here for managing beans.
Like if getBean could be SugarBean|bool, just imagine if returns boolean.....

Comment (samus-aran) [reply lib/API/v8/Controller/ModuleController.php]: Fair point that there needs to be an exercise to cover more ground for exceptions, but that isn't required here as that would be need to be tackled incrementally. How long is a piece of string would be the phrase I would use for this task.

**Comment by Dillon-Brown** [comment]
Assessed

**Pull request merged**